sdk/storage/Azure.Storage.DataMovement.Blobs/src/BlobContainerClientExtensions.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading.Tasks;
+using Azure.Storage.DataMovement;
+using Azure.Storage.DataMovement.Blobs;
+
+namespace Azure.Storage.Blobs
+{
+    /// <summary>
+    /// Data movement extension methods for the <see cref="Azure.Storage.Blobs.BlobContainerClient"/>.
+    /// </summary>
+    public static class BlobContainerClientExtensions
+    {
+        private static Lazy<TransferManager> s_defaultTransferManager = new Lazy<TransferManager>();
+
+        /// <summary>
+        /// Uploads the entire contents of local directory to the blob container.
+        /// </summary>
+        /// <param name="client">The <see cref="BlobContainerClient"/> used for service operations.</param>
+        /// <param name="localDirectoryPath">The full path to the local directory to be uploaded.</param>
+        /// <param name="blobDirectoryPrefix">Optionally specifies the directory prefix to be prepended to all uploaded files.</param>
+        /// <param name="options">Options which control the directory upload.</param>
+        /// <returns>A <see cref="DataTransfer"/> instance which can be used track progress and wait for completion with <see cref="DataTransfer.AwaitCompletion"/>.</returns>
+        public static async Task<DataTransfer> StartUploadDirectoryAsync(this BlobContainerClient client, string localDirectoryPath, string blobDirectoryPrefix = default, BlobContainerClientTransferOptions options = default)
```

Comment:
Feels like `blobDirectoryPrefix` should be in the options bag. Storage aggressively tries to avoid optional parameters in the main signature.

Reply:
We wanted the `blobDirectoryPrefix` to not go in the options bag that way customers are well aware that they can upload to a specific directory instead of only at the root of the container.

I think the reasoning behind it is cause potentially customers could upload to the root of the container by accident(?) @schaabs

Reply:
I've updated this to pull apart into two overloads a simplified one with no options parameter, where `blobDirectoryPrefix` is optional.

```c#
public static Task<DataTransfer> StartUploadDirectoryAsync(this BlobContainerClient client, string localDirectoryPath, string blobDirectoryPrefix = default)
```

Then I included `BlobDirectoryPrefix` on `BlobContainerClientTransferOptions`, and added the second overload which only takes `localDirectoryPath` and `options`, both required parameters.

```c#
public static async Task<DataTransfer> StartUploadDirectoryAsync(this BlobContainerClient client, string localDirectoryPath, BlobContainerClientTransferOptions options)
```

I think this simplifies the use of optional arguments substantially as it eliminates having two optional parameters in the same overload, while still having a convenience method where it's possible just to specify the source and, optionally, target directory. However, if we really wanted to avoid optional parameters all together we could split the first overload into to, resulting in the three overloads.

```c#
public static Task<DataTransfer> StartUploadDirectoryAsync(this BlobContainerClient client, string localDirectoryPath)
public static Task<DataTransfer> StartUploadDirectoryAsync(this BlobContainerClient client, string localDirectoryPath, string blobDirectoryPrefix)
public static async Task<DataTransfer> StartUploadDirectoryAsync(this BlobContainerClient client, string localDirectoryPath, BlobContainerClientTransferOptions options)
```

Thoughts?

Reply:
Looking over the API changes, this looks good. This is currently what's updated in the PR right? I'm going to resolve this convo.